mwan3: speedups (1230ms to 41ms)

champtar opened the pull request:
## 📦 Package Details

**Maintainer:** @feckert 

**Description:**
From 1230ms to 41ms

## 🧪 Run Testing Details

- **OpenWrt Version:** 24.10 (with all mwan3 master commits backported)
- **OpenWrt Target/Subtarget:**
- **OpenWrt Device:**

champtar commented:
@pschmitt @brianjmurrell @wrsolichin as you contributed to mwan3 in the past, looking for testers for this change.
I'm already dogfooding my changes, and i've tried my best to make the change transparent, but would enjoy some testing

pschmitt commented:
What's wrong with `cat`?

champtar commented:
> What's wrong with `cat`?

It's just slow. Removing `$(cat ...)` and `%(...)` we go from 1230ms to 485ms, fork/exec are pretty expensive just to read 2 chars

feckert commented:
@champtar Thank you for this PR. It will take a while to review these changes. I'll see when I have the time.

feckert commented on net/mwan3/files/usr/share/rpcd/ucode/mwan3:
When I look at the other Ucode implementations in the repo, they all use `tabs` for indentation.
So can you please change that.

I still have to look into ucode. So I can't say much about it.
But I welcome the fact that we are now using ucode instead of shell :+1:

feckert commented on net/mwan3/files/usr/sbin/mwan3track:
Is the `jobs` a build in from the shell?
If I understand this correctly, then I get all PIDs from the child jobs of my process with `jobs -p` and not the job IDs. So I can call the `kill` command with the results from `jobs -p`?

Can you also write in the commit message why you removed `TRACK_PID` `SLEEP_PID`
`Just use 'job -p' to list the subprocesses currently running, so we do not needed the global variables TRACK_PID and SLEEP_PID  anymore`

champtar replied:
`jobs` is a builtin yes, `jobs -p` returns the process IDs, not the job IDs, so we can use kill directly
I've improved the commit message

feckert commented on net/mwan3/files/lib/mwan3/common.sh:
Can you please move the add of the readfile function to a separate commit.
As a preparatory commit.
Then the commit message is clearer and you can focus on the changes in this commit.

```
mwan3: add new readfile function to speedup variable read from file

This is a preparation  for adding an utility function readfile() to read
directly to a variable. Using '$(cat ...)' causes a fork/execve which is
extremely slow just to read '~2' chars.
```

champtar replied:
done

feckert reviewed (changes requested):
Nice work :+1:
LGTM please take a look at my comments.

systemcrash commented on net/mwan3/files/usr/share/rpcd/ucode/mwan3:
```diff
@@ -0,0 +1,201 @@
+            'up': netstatus.up,
+            'track_ip': track_ips,
+        };
+    });
```
`uci.unload('mwan3');`

champtar replied:
I'm creating the cursor at the top of the function, so I don't think I need to manually unload() and I can just leave it to the GC to free memory
Only example where I see unload is in `ddns.uc` but the cursor is global

systemcrash commented on net/mwan3/files/usr/share/rpcd/ucode/mwan3:
```diff
@@ -0,0 +1,201 @@
+        if (request.args.interface != null && request.args.interface != ifname) {
+            return;
+        }
+        let netstatus = ubus.call(sprintf('network.interface.%s', ifname), 'status', {});
+        let uptime = get_uptime();
+        let uci_track_ips = intf['track_ip'];
+        let track_status = get_mwan3track_status(ifname, uci_track_ips, procd);
+        let track_ips = [];
```
const

champtar replied:
done

systemcrash commented on net/mwan3/files/usr/share/rpcd/ucode/mwan3:
```diff
@@ -0,0 +1,201 @@
+
+function get_policies(version) {
+    const ipt = popen(policies_cmd[version], 'r');
+    let policies = {};
```
const

systemcrash commented on net/mwan3/files/usr/share/rpcd/ucode/mwan3:
```diff
@@ -0,0 +1,201 @@
+        return 'disabled';
+    }
+    if (procd?.[sprintf('track_%s', iface)]?.running) {
+        let started = get_str(iface, 'STARTED');
```
const

champtar replied:
done

systemcrash commented on net/mwan3/files/usr/share/rpcd/ucode/mwan3:
```diff
@@ -0,0 +1,201 @@
+    const ipset = popen(sprintf('ipset -o save list mwan3_connected_ipv%s', version), 'r');
+    let ips = [];
+    for (let line = ipset.read('line'); length(line); line = ipset.read('line')) {
+        let m = match(line, ipset_save_re);
```
const

champtar replied:
done

systemcrash commented on net/mwan3/files/usr/share/rpcd/ucode/mwan3:
```diff
@@ -0,0 +1,201 @@
+        return [];
+    }
+    const ipset = popen(sprintf('ipset -o save list mwan3_connected_ipv%s', version), 'r');
+    let ips = [];
```
const?

systemcrash commented on net/mwan3/files/usr/share/rpcd/ucode/mwan3:
```diff
@@ -0,0 +1,201 @@
+let ubus = require('ubus').connect();
+
+function get_str_raw(iface, property) {
+    return readfile(sprintf('/var/run/mwan3track/%s/%s', iface, property));
```
I get the sneaking suspicion that this might already be available via `procd` and something like `ubus call service list`. Altho this way may be 'less expensive' to call.

champtar replied:
I don't understand your comment here.
mwan3track store its state in /var/run/mwan3track/, here we read those small files to export all the info via rpcd/ubus

systemcrash replied:
I was thinking you could already read it from ubus, like `ubus call network.interface.mwan3 status` or something.

champtar replied:
I already call ubus twice, many of those property are mwan3track specific and must be read from `/var/run/mwan3track/`

champtar commented on net/mwan3/files/usr/share/rpcd/ucode/mwan3:
```diff
@@ -0,0 +1,201 @@
+    let ips = [];
+    for (let line = ipset.read('line'); length(line); line = ipset.read('line')) {
+        let m = match(line, ipset_save_re);
+        if (length(m) == 2) {
```
This is ucode (https://ucode.mein.io/), not js

length doesn't error out, it just 'Returns null if the given argument is not an object, array or string.'
https://ucode.mein.io/module-core.html#length

```
# ucode -p "length(null) == 2"
false
# ucode -p "length(null)"
null
# ucode -p "length([])"
0
```

champtar commented on net/mwan3/files/lib/mwan3/common.sh:
```diff
@@ -86,30 +86,48 @@ mwan3_get_src_ip()
+		return
+	fi
+	readfile CMDLINE /proc/$PID/cmdline 2>/dev/null
+	if [ $CMDLINE != "/bin/sh/usr/sbin/mwan3track$INTERFACE" ]; then
```
comparison works just fine, and suggested change wouldn't even work

champtar commented on net/mwan3/files/usr/share/rpcd/ucode/mwan3:
```diff
@@ -0,0 +1,201 @@
+function interfaces_status(request) {
+    const uci = cursor();
+    let procd = ubus.call('service', 'list', { 'name': 'mwan3' })?.mwan3?.instances;
+    let interfaces = {};
```
done

champtar commented:
@feckert thanks for the review, I've added a dependency on rpcd-mod-ucode, tell me if this is what you wanted
(we can also split this in a second package)
@systemcrash replaced let with const as much as I could, but there are 2 open comments

brianjmurrell commented on net/mwan3/files/lib/mwan3/common.sh:
```diff
@@ -94,28 +94,40 @@ readfile() {
+		return
+	fi
+	readfile CMDLINE /proc/$PID/cmdline 2>/dev/null
+	if [ $CMDLINE != "/bin/sh/usr/sbin/mwan3track$INTERFACE" ]; then
```
> change line to a valid shell script
> 
> ```
> if [ $CMDLINE != "/bin/sh/usr/sbin/mwan3track$INTERFACE" ]; then
> ```

FWIW, that is not invalid.  It's syntactically correct.  `{}` are optional in variables unless you need the closing `}` to delineate the variable ma,e in such a case as as `${INTERFACE}_foo`.

Otherwise always using `{}` in shell variables is a style issue only.

champtar commented:
@feckert I'll let you do a last review, please don't merge right away I haven't run tested yet the last changes

champtar commented on net/mwan3/files/usr/share/rpcd/ucode/mwan3:
```diff
@@ -0,0 +1,201 @@
+const policies_re = regexp('^-A mwan3_policy_([^ ]+) .*?--comment "([^"]+)"');
+
+function get_policies(version) {
+	const ipt = popen(policies_cmd[version], 'r');
```
Yes I do, fixed

champtar commented on net/mwan3/files/usr/share/rpcd/ucode/mwan3:
```diff
@@ -0,0 +1,201 @@
+		for (let ip in uci_track_ips) {
+			push(track_ips, {
+				'ip': ip,
+				'status': get_str(ifname, sprintf('TRACK_%s', ip)),
```
Done

champtar commented on net/mwan3/files/usr/share/rpcd/ucode/mwan3:
```diff
@@ -0,0 +1,201 @@
+			'age': get_x_time(uptime, ifname, 'TIME'),
+			'online': get_x_time(uptime, ifname, 'ONLINE'),
+			'offline': get_x_time(uptime, ifname, 'OFFLINE'),
+			'uptime': netstatus.uptime,
```
Done

champtar commented:
@feckert @systemcrash thanks again for the reviews, all fixed
